components/net/filemanager_thread.rs

```diff
@@ -0,0 +1,145 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use ipc_channel::ipc::{self, IpcReceiver, IpcSender};
+use net_traits::filemanager_thread::{FileManagerThreadMsg, FileManagerResult, FileManagerThreadError};
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use util::thread::spawn_named;
+use uuid::Uuid;
+
+pub struct FileManager {
+    receiver: IpcReceiver<FileManagerThreadMsg>,
+    idmap: RefCell<HashMap<Uuid, String>>,
+}
+
+enum SelectedPort {
+    Single(IpcSender<FileManagerResult<(Uuid, PathBuf, u64)>>),
+    Multiple(IpcSender<FileManagerResult<Vec<(Uuid, PathBuf, u64)>>>)
+}
+
+impl FileManager {
+    fn new(recv: IpcReceiver<FileManagerThreadMsg>) -> FileManager {
+        FileManager {
+            receiver: recv,
+            idmap: RefCell::new(HashMap::new()),
+        }
+    }
+
+    fn new_thread() -> IpcSender<FileManagerThreadMsg> {
+        let (chan, recv) = ipc::channel().unwrap();
+
+        spawn_named("FileManager".to_owned(), move || {
+            FileManager::new(recv).start();
+        });
+
+        chan
+    }
+
+    /// Start the file manager event loop
+    pub fn start(&mut self) {
+        loop {
+            match self.receiver.recv().unwrap() {
+                FileManagerThreadMsg::SelectFile(sender) => self.select(SelectedPort::Single(sender)),
+                FileManagerThreadMsg::SelectFiles(sender) => self.select(SelectedPort::Multiple(sender)),
+                FileManagerThreadMsg::ReadFile(sender, id) => self.read_file(sender, id),
+                FileManagerThreadMsg::DeleteFileID(id) => self.delete_fileid(id),
+            }
+        }
+    }
+}
+
+impl FileManager {
+    fn select(&mut self, sender: SelectedPort) {
+        // TODO: Pull the dialog UI in and get selected
+        match sender {
```

Comment:
I sort of feel that this would be better as two functions with no SelectedPort, and have the dialog code take an argument specifying single or multiple.